### DISCUSSION TITLE
Variable prefixes and crossref

### DISCUSSION THREAD

**Discussion by lou1306** [eclipse-langium/langium#1316] [Q&A] [answered]
I am writing a Langium grammar for a DSL where some variables ("properties") need to be prefixed with an `@` when referenced. (Something similar to Bash or Php, where `$var` is a reference to `var`).

My desiderata are:
1. The grammar should accept `@var` but reject `@ var`
2. Langium should be able to cross-ref `var` to its declaration.

Right now I have
```
{infer PropRef} '@' variable=[Prop]
```

This satisfies 2. but not 1 (we can put whitespace after the `@` sign and the parser will not complain). If I instead write a terminal rule, such as

```
terminal PROPREF: /@([_a-zA-Z][\w_]*)/;
```

Then I can achieve 1. but not 2. because Langium looks for `@var` instead of `var`.

Is there an easy way to make it work? Ideally I would like to have a terminal rule that only returns part of the whole match (e.g., a group).

**Comment by msujew** [comment, marked as answer]
Hey @lou1306,

you're already doing well by using `PROPREF` as part of the reference. I.e. by using `variable=[Prop:PROPREF]`.

For the second part, this becomes a bit less trivial. Langium exposes an AST element always by its canonical name. By default, this is the name that you assign using the `name` AST property. I.e. if you name an element `var`, it will only be addressable by the name `var`. If you want to give it a different (or even multiple) canonical names, you will need to modify the `ScopeComputation` service. We have a guide on how to do that [here](https://langium.org/guides/scoping/qualified-name/). The guide uses fully qualified name scoping, but that is also just another way of exposing an element using a different name (a fully qualified one).

**Reply by lou1306** [reply]
That worked! Thank you for pointing me in the right direction.
